File: src/humanize/failure.rs
//! Humanized failure recovery — how a "human" would handle action failures

use rand::Rng;

use super::config::{FailureStyle, HumanizationConfig};

/// Error codes that can occur during browser action execution
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub enum ActionErrorCode {
    /// Target element not found on the page
    ElementNotFound,
    /// Click action failed
    ClickFailed,
    /// Navigation timed out
    NavigationTimeout,
    /// Proxy is unresponsive
    ProxyDead,
    /// Fingerprint was detected / rejected
    FingerprintRejected,
    /// Browser session crashed
    SessionCrashed,
    /// Rate limited by target site
    RateLimited,
    /// Content visibility blocked
    ContentBlinded,
    /// Unknown / unexpected error
    Unknown,
}

impl ActionErrorCode {
    pub fn as_str(&self) -> &'static str {
        match self {
            ActionErrorCode::ElementNotFound => "ELEMENT_NOT_FOUND",
            ActionErrorCode::ClickFailed => "CLICK_FAILED",
            ActionErrorCode::NavigationTimeout => "NAVIGATION_TIMEOUT",
            ActionErrorCode::ProxyDead => "PROXY_DEAD",
            ActionErrorCode::FingerprintRejected => "FINGERPRINT_REJECTED",
            ActionErrorCode::SessionCrashed => "SESSION_CRASHED",
            ActionErrorCode::RateLimited => "RATE_LIMITED",
            ActionErrorCode::ContentBlinded => "CONTENT_BLINDED",
            ActionErrorCode::Unknown => "UNKNOWN",
        }
    }
}

/// Recovery action the system should take
#[derive(Debug, Clone)]
pub enum RecoveryAction {
    /// Retry immediately
    Retry,
    /// Wait then retry
    RetryAfter { wait_ms: u32 },
    /// Give up on this action
    GiveUp,
    /// Switch to a different approach (e.g., different selector)
    SwitchApproach,
    /// Skip this action and continue
    Skip,
}

/// Decision on whether and how to retry after a failure
#[derive(Debug, Clone)]
pub struct RetryDecision {
    pub action: RecoveryAction,
    /// Probability that this decision was correct (for logging/analytics)
    pub estimated_success_probability: f32,
}

impl RetryDecision {
    pub fn retry() -> Self {
        Self {
            action: RecoveryAction::Retry,
            estimated_success_probability: 0.6,
        }
    }

    pub fn retry_after(wait_ms: u32) -> Self {
        Self {
            action: RecoveryAction::RetryAfter { wait_ms },
            estimated_success_probability: 0.75,
        }
    }

    pub fn give_up() -> Self {
        Self {
            action: RecoveryAction::GiveUp,
            estimated_success_probability: 0.0,
        }
    }

    pub fn switch_approach() -> Self {
        Self {
            action: RecoveryAction::SwitchApproach,
            estimated_success_probability: 0.5,
        }
    }

    pub fn skip() -> Self {
        Self {
            action: RecoveryAction::Skip,
            estimated_success_probability: 0.0,
        }
    }
}

/// Humanized retry decision based on error type, attempt number, and config
pub fn humanized_retry_decision(
    error: ActionErrorCode,
    attempt: u32,
    config: &HumanizationConfig,
) -> RetryDecision {
    let style = &config.failure;

    match style {
        FailureStyle::Instant => {
            // Machine-style: always retry up to limit
            if attempt < 3 {
                RetryDecision::retry()
            } else {
                RetryDecision::give_up()
            }
        }
        FailureStyle::Human {
            min_wait_ms,
            max_wait_ms,
            max_retries,
            give_up_chance,
        } => {
            let mut rng = rand::thread_rng();

            // Give up immediately on certain fatal errors
            if matches!(
                error,
                ActionErrorCode::SessionCrashed
                    | ActionErrorCode::FingerprintRejected
                    | ActionErrorCode::ProxyDead
            ) {
                if attempt >= 1 {
                    // Fatal errors: give up after 1 attempt (or occasionally retry proxy once)
                    if error == ActionErrorCode::ProxyDead && attempt == 1 {
                        let wait = rng.gen_range(*min_wait_ms..=*max_wait_ms);
                        return RetryDecision::retry_after(wait);
                    }
                    return RetryDecision::give_up();
                }
            }

            // Already exhausted retries
            if attempt >= *max_retries {
                return RetryDecision::give_up();
            }

            // Give up with some probability even before max retries
            if rng.gen::<f32>() < *give_up_chance {
                return RetryDecision::give_up();
            }

            // Human hesitation: wait before retry
            let wait_ms = rng.gen_range(*min_wait_ms..=*max_wait_ms);

            // More attempts = longer waits (less motivated to keep trying)
            let attempts_multiplier = 1.0 + (attempt as f32 * 0.3);
            let adjusted_wait = (wait_ms as f32 * attempts_multiplier) as u32;

            RetryDecision::retry_after(adjusted_wait)
        }
    }
}

/// Get suggested alternative approach for a given error code
pub fn suggested_approach(error: ActionErrorCode) -> &'static str {
    match error {
        ActionErrorCode::ElementNotFound => "retry_with_wait",
        ActionErrorCode::ClickFailed => "use_offset_click",
        ActionErrorCode::NavigationTimeout => "retry_with_different_proxy",
        ActionErrorCode::ProxyDead => "switch_proxy",
        ActionErrorCode::FingerprintRejected => "switch_fingerprint",
        ActionErrorCode::SessionCrashed => "restart_session",
        ActionErrorCode::RateLimited => "wait_longer",
        ActionErrorCode::ContentBlinded => "scroll_into_view",
        ActionErrorCode::Unknown => "skip",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn medium_config() -> HumanizationConfig {
        HumanizationConfig::from_level(super::super::config::HumanizationLevel::Medium)
    }

    #[test]
    fn test_fatal_error_gives_up_quickly() {
        let config = medium_config();
        let decision = humanized_retry_decision(ActionErrorCode::SessionCrashed, 1, &config);
        assert!(matches!(decision.action, RecoveryAction::GiveUp));
    }

    #[test]
    fn test_non_fatal_error_retries() {
        let config = medium_config();
        let decision = humanized_retry_decision(ActionErrorCode::ElementNotFound, 0, &config);
        assert!(matches!(
            decision.action,
            RecoveryAction::RetryAfter { .. } | RecoveryAction::Retry
        ));
    }

    #[test]
    fn test_give_up_chance_respected() {
        let config = HumanizationConfig::from_level(super::super::config::HumanizationLevel::High);
        let give_up_chance = match &config.failure {
            FailureStyle::Human { give_up_chance, .. } => *give_up_chance,
            _ => panic!("expected Human style"),
        };
        assert!(
            give_up_chance > 0.1,
            "High level should have meaningful give_up_chance"
        );
    }
}

File: backend/internal/behavior/humanize/failure.go
package humanize

import "math/rand/v2"

// ActionErrorCode enumerates known failure modes for browser actions.
type ActionErrorCode int

const (
	ErrElementNotFound     ActionErrorCode = 0
	ErrClickFailed         ActionErrorCode = 1
	ErrNavigationTimeout   ActionErrorCode = 2
	ErrProxyDead           ActionErrorCode = 3
	ErrFingerprintRejected ActionErrorCode = 4
	ErrSessionCrashed      ActionErrorCode = 5
	ErrRateLimited         ActionErrorCode = 6
	ErrContentBlinded      ActionErrorCode = 7
	ErrUnknown             ActionErrorCode = 8
)

func (e ActionErrorCode) String() string {
	switch e {
	case ErrElementNotFound:
		return "ELEMENT_NOT_FOUND"
	case ErrClickFailed:
		return "CLICK_FAILED"
	case ErrNavigationTimeout:
		return "NAVIGATION_TIMEOUT"
	case ErrProxyDead:
		return "PROXY_DEAD"
	case ErrFingerprintRejected:
		return "FINGERPRINT_REJECTED"
	case ErrSessionCrashed:
		return "SESSION_CRASHED"
	case ErrRateLimited:
		return "RATE_LIMITED"
	case ErrContentBlinded:
		return "CONTENT_BLINDED"
	default:
		return "UNKNOWN"
	}
}

// RecoveryActionType enumerates what to do after a failure.
type RecoveryActionType int

const (
	RecoveryRetry          RecoveryActionType = 0
	RecoveryRetryAfter     RecoveryActionType = 1 // wait, then retry
	RecoveryGiveUp         RecoveryActionType = 2
	RecoverySwitchApproach RecoveryActionType = 3
	RecoverySkip           RecoveryActionType = 4
)

// RecoveryAction describes the recovery strategy after a failure.
type RecoveryAction struct {
	Type   RecoveryActionType
	WaitMs uint32 // only for RetryAfter
}

// RetryDecision combines a recovery action with an estimated success probability.
type RetryDecision struct {
	Action                      RecoveryAction
	EstimatedSuccessProbability float64
}

// Convenience constructors
func retryNow() RetryDecision {
	return RetryDecision{Action: RecoveryAction{Type: RecoveryRetry}, EstimatedSuccessProbability: 0.6}
}
func retryAfter(ms uint32) RetryDecision {
	return RetryDecision{Action: RecoveryAction{Type: RecoveryRetryAfter, WaitMs: ms}, EstimatedSuccessProbability: 0.75}
}
func giveUp() RetryDecision {
	return RetryDecision{Action: RecoveryAction{Type: RecoveryGiveUp}, EstimatedSuccessProbability: 0.0}
}
func switchApproach() RetryDecision {
	return RetryDecision{Action: RecoveryAction{Type: RecoverySwitchApproach}, EstimatedSuccessProbability: 0.5}
}
func skipRetry() RetryDecision {
	return RetryDecision{Action: RecoveryAction{Type: RecoverySkip}, EstimatedSuccessProbability: 0.0}
}

// HumanizedRetryDecision decides what to do after an action fails.
// Incorporates human-like hesitation, fatal error recognition, and early give-up.
func HumanizedRetryDecision(errCode ActionErrorCode, attempt uint32, config *HumanizationConfig) RetryDecision {
	switch config.Failure.Type {
	case FailureInstant:
		// Machine-like: always retry up to max retries
		if attempt < config.Failure.MaxRetries {
			return retryNow()
		}
		return giveUp()

	case FailureHuman:
		// 1. Fatal errors: give up immediately (with one exception for ProxyDead)
		if isFatal(errCode) {
			if errCode == ErrProxyDead && attempt <= 1 {
				wait := randRangeUint32(config.Failure.MinWaitMs, config.Failure.MaxWaitMs)
				return retryAfter(wait)
			}
			return giveUp()
		}

		// 2. Exhausted retries
		if attempt >= config.Failure.MaxRetries {
			return giveUp()
		}

		// 3. Random early give-up (humans sometimes just quit)
		if rand.Float64() < config.Failure.GiveUpChance {
			return giveUp()
		}

		// 4. Human hesitation: wait, then retry
		waitMs := randRangeUint32(config.Failure.MinWaitMs, config.Failure.MaxWaitMs)
		multiplier := 1.0 + float64(attempt)*0.3
		adjustedWait := uint32(float64(waitMs) * multiplier)
		return retryAfter(adjustedWait)

	default:
		return giveUp()
	}
}

// SuggestedApproach returns a human-readable suggestion for how to recover from an error.
func SuggestedApproach(errCode ActionErrorCode) string {
	switch errCode {
	case ErrElementNotFound:
		return "retry_with_wait"
	case ErrClickFailed:
		return "use_offset_click"
	case ErrNavigationTimeout:
		return "retry_with_different_proxy"
	case ErrProxyDead:
		return "switch_proxy"
	case ErrFingerprintRejected:
		return "switch_fingerprint"
	case ErrSessionCrashed:
		return "restart_session"
	case ErrRateLimited:
		return "wait_longer"
	case ErrContentBlinded:
		return "scroll_into_view"
	default:
		return "skip"
	}
}

func isFatal(errCode ActionErrorCode) bool {
	switch errCode {
	case ErrSessionCrashed, ErrFingerprintRejected, ErrProxyDead:
		return true
	default:
		return false
	}
}
